flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/BatchJobRecoveryHandler.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.flink.runtime.executiongraph.ResultPartitionBytes;
+import org.apache.flink.runtime.executiongraph.TaskExecutionStateTransition;
+import org.apache.flink.runtime.executiongraph.failover.FailoverStrategy;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.jobgraph.IntermediateResultPartitionID;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.shuffle.ShuffleMaster;
+import org.apache.flink.util.function.ConsumerWithException;
+import org.apache.flink.util.function.QuadConsumerWithException;
+import org.apache.flink.util.function.TriConsumer;
+
+import org.slf4j.Logger;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+/** Interface for handling batch job recovery. */
+public interface BatchJobRecoveryHandler {
+
+    /**
+     * Stops the job recovery handler and optionally clears up.
+     *
+     * @param clearUp whether to clear up.
+     */
+    void stop(boolean clearUp);
+
+    /**
+     * Starts the recovery process and sets up listeners for recovery completion or failure.
+     *
+     * @param recoverFinishedListener a listener called in case recovery finished.
+     * @param recoverFailedListener a runnable called in case recovery fails.
+     */
+    void startRecovering(
+            Consumer<Set<JobVertexID>> recoverFinishedListener, Runnable recoverFailedListener);
+
+    /** Determines whether the job needs to undergo recovery. */
+    boolean needRecover();
+
+    /** Determines whether the job is recovering. */
+    boolean isRecovering();
+
+    /**
+     * Notifies that a set of execution vertices have been reset.
+     *
+     * @param vertices a collection of execution vertex IDs that have been reset.
+     */
+    void notifyExecutionVertexReset(Collection<ExecutionVertexID> vertices);
+
+    /**
+     * Notifies the initialization of a job vertex.
+     *
+     * @param jobVertexId the ID of the job vertex being initialized.
+     * @param parallelism the parallelism degree of the job vertex.
+     * @param jobVertexInputInfos a map of intermediate dataset IDs to job vertex input info.
+     */
+    void notifyExecutionJobVertexInitialization(
+            JobVertexID jobVertexId,
+            int parallelism,
+            Map<IntermediateDataSetID, JobVertexInputInfo> jobVertexInputInfos);
+
+    /** Notifies that an execution has finished. */
+    void notifyExecutionFinished(
+            ExecutionVertexID executionVertexId, TaskExecutionStateTransition taskExecutionState);
+
+    /** Initializes the recovery handler with the necessary components. */
+    void initialize(
+            Logger log,
+            ExecutionGraph executionGraph,
+            ShuffleMaster<?> shuffleMaster,
```

Comment:
It's better to introduce a context to host all the methods needed for recovery.
This can help the needed method to be well documented instead using `Consumer/QuadConsumerWithException/Runnable/...`.